pkg/virtual/syncer/coordination/helpers.go

```diff
@@ -0,0 +1,228 @@
+		return syncTargetKey == key
+	})
+}
+
+type Object interface {
+	metav1.Object
+	runtime.Object
+}
+
+func UpstreamViewChanged(old, new Object, equality func(old, new interface{}) bool) bool {
+	old = old.DeepCopyObject().(Object)
+	new = new.DeepCopyObject().(Object)
+
+	oldAnnotations := make(map[string]string)
+	for name, value := range old.GetAnnotations() {
+		if !strings.HasPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix) {
+			oldAnnotations[name] = value
+		}
+	}
+	old.SetAnnotations(oldAnnotations)
+
+	newAnnotations := make(map[string]string)
+	for name, value := range new.GetAnnotations() {
+		if strings.HasPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix) {
+			newAnnotations[name] = value
+		}
+	}
+	new.SetAnnotations(newAnnotations)
+
+	return !equality(old, new)
+}
+
+type SyncerViewManager[T Object] interface {
+	GetFilteredSyncerViews(ctx context.Context, gvr schema.GroupVersionResource, upstreamResource T, keepSyncTarget func(key string) bool) (map[string]T, error)
+	GetSyncerViews(ctx context.Context, gvr schema.GroupVersionResource, upstreamResource T) (map[string]T, error)
+	GetSyncerView(ctx context.Context, gvr schema.GroupVersionResource, upstreamResource T, syncTargetKey string) (T, error)
+}
+
+var _ SyncerViewManager[Object] = (*syncerViewManager[Object])(nil)
+
+type syncerViewManager[T Object] struct {
+	transformations.SyncerResourceTransformer
+}
+
+func NewSyncerViewManager[T Object](transformationProvider transformations.TransformationProvider,
+	summarizingRulesprovider transformations.SummarizingRulesProvider) SyncerViewManager[T] {
+	return &syncerViewManager[T]{
+		transformations.SyncerResourceTransformer{
+			TransformationProvider:   transformationProvider,
+			SummarizingRulesProvider: summarizingRulesprovider,
+		},
+	}
+}
+
+func NewDefaultSyncerViewManager[T Object]() SyncerViewManager[T] {
+	return NewSyncerViewManager[T](&transformations.SpecDiffTransformation{}, &transformations.DefaultSummarizingRules{})
+}
+
+func toUnstructured[T Object](obj T) (*unstructured.Unstructured, error) {
+	unstructured := &unstructured.Unstructured{Object: map[string]interface{}{}}
+	raw, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	unstructured.Object = raw
+	return unstructured, nil
+}
+
+func fromUnstructured[T Object](unstr *unstructured.Unstructured) (T, error) {
+	var obj T
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(unstr.Object, &obj)
+	if err != nil {
+		return obj, err
+	}
+	return obj, nil
+}
+
+func (svm *syncerViewManager[T]) GetFilteredSyncerViews(ctx context.Context, gvr schema.GroupVersionResource, upstreamResource T, keepSyncTarget func(key string) bool) (map[string]T, error) {
+	syncerViews := make(map[string]T)
+	for name := range upstreamResource.GetAnnotations() {
+		if strings.HasPrefix(name, v1alpha1.InternalSyncerViewAnnotationPrefix) {
```

Comment:
continue early. Remove indention for the rest.

Reply:
Done.